clusterctl/validation/validate_cluster_api_objects.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+	"io"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/common"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+	"sigs.k8s.io/cluster-api/pkg/controller/noderefutil"
+)
+
+func ValidateClusterAPIObjects(w io.Writer, clusterApiClient *clientset.Clientset, k8sClient kubernetes.Interface, clusterName string, namespace string) error {
+	fmt.Fprintf(w, "Validating cluster API objects in namespace %q\n", namespace)
+
+	cluster, err := getClusterObject(clusterApiClient, clusterName, namespace)
+	if err != nil {
+		return err
+	}
+	if err := validateClusterObject(w, cluster); err != nil {
```

Comment:
Seems like the cluster object you get above is not used anywhere else in this method. I recommend you `getClusterObject` in `validateClusterObject` itself.

Reply:
Separating them makes testing much easier. So when testing ```validateClusterObject```, I can just provide fake cluster objects. If I move ```getClusterObject``` into ```validateClusterObject```, I will have to manipulate with the fake client, too.

To keep current tests, another way is to change to the following:
```
func ValidateClusterAPIObjects(...) {
    getAndValidateClusterObject(...)
    getAndValidateMachineObjects(...)
}
func getAndValidateClusterObject(...) {
    cluster := getClusterObject(...)
    validateClusterObject(cluster ...)
}
func getAndValidateMachineObjects(...) {
    machines := getMachineObjects(...)
    validateMachineObjects(machines ...)
}
```
But I think the above is actually making current code more complex.